src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,137 @@ private int SkipBlockComment(int i)
+            }
+
+            int maxNumberOfHexDigits = (isBracketedSequence ? 6 : 4);
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            for (int i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Bracketed sequence has been terminated.
+                if (isBracketedSequence && (c == '}'))
+                {
+                    if (i == 0)
+                    {
+                        // Variable sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidBracketedUnicodeEscapeSequence);
+                        return s_unknownChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    if (isBracketedSequence)
+                    {
+                        ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    }
+                    else
+                    {
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                    }
+
+                    return s_unknownChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (isBracketedSequence && !isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    return s_unknownChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
+            }
+            else
+            {
+                // Place the error indicator under only the hex digits in the esc sequence.
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex + 3, _currentIndex - 1);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequenceValue);
+                return s_unknownChar;
             }
         }
 
+        private bool IsSurrogatePair(string str, ref char nonSurrogatePairChar, StringBuilder sb1 = null, StringBuilder sb2 = null) 
```

Comment:
If we use overloads then presumably the version that takes one SB will delegate to the overload that takes two SBs (and has the implementation).  And for that delegation, it would pass null for the second SB - at least that is how I've done overloads before.  Are you thinking of a different way?  

BTW this method is perhaps a bit more general purpose than it needs to be since the only existing callers pass either one or two SBs.  There are no places that call it without any SB.  So we could `Diag.Assert` that sb1 is never null and do away with the first `sb1 != null` check - **if** you're comfortable with that.  But we would still need to check `sb2 != null` since sb2 would be null if we were called from the overload that takes just sb1.